common/numerated/src/interval.rs

```diff
@@ -85,352 +67,247 @@ impl<T: Numerated> NonEmptyInterval<T> {
+            Some(start) => {
+                let end = T::max_value();
+                debug_assert!(start <= end, "`T: UpperBounded` impl error");
+                Ok(Self { start, end })
             }
+            None => Err(EmptyRangeError),
         }
     }
 }
 
-impl<T: Numerated, I: Into<T::B>> TryFrom<Range<I>> for Interval<T> {
-    type Error = IntoIntervalError;
+impl<T: Numerated + LowerBounded + UpperBounded, I: Into<T::Bound>> TryFrom<RangeTo<I>>
+    for Interval<T>
+{
+    type Error = EmptyRangeError;
 
-    fn try_from(range: Range<I>) -> Result<Self, Self::Error> {
-        Self::try_from((range.start, range.end))
+    fn try_from(range: RangeTo<I>) -> Result<Self, Self::Error> {
+        let end: T::Bound = range.end.into();
+
+        let Some(end) = end.unbound() else {
+            return Ok(Self::from(..));
+        };
+
+        let start = T::min_value();
+        end.dec_if_gt(start)
+            .map(|end| {
+                debug_assert!(start <= end, "`T: LowerBounded` impl error");
+                Self { start, end }
+            })
+            .ok_or(EmptyRangeError)
     }
 }
 
 impl<T: Numerated> TryFrom<RangeInclusive<T>> for Interval<T> {
-    type Error = IntoIntervalError;
+    type Error = IncorrectRangeError;
 
     fn try_from(range: RangeInclusive<T>) -> Result<Self, Self::Error> {
         let (start, end) = range.into_inner();
-        NonEmptyInterval::new(start, end)
-            .ok_or(IntoIntervalError)
-            .map(Into::into)
+        (start <= end)
+            .then_some(Self { start, end })
+            .ok_or(IncorrectRangeError)
     }
 }
 
-impl<T: Numerated> NonEmptyInterval<T> {
-    /// Returns amount of elements in interval in `T::N` if it's possible.
-    /// None means that interval size is bigger, than `T::N::max_value()`.
-    pub fn raw_size(&self) -> Option<T::N> {
-        let (start, end) = self.into_inner();
+impl<T, S, E> TryFrom<(S, E)> for Interval<T>
+where
+    T: Numerated + UpperBounded,
+    S: Into<T::Bound>,
+    E: Into<T::Bound>,
+{
+    type Error = TryFromRangeError;
 
-        // guarantied by `Self`
-        debug_assert!(start <= end);
-
-        let distance = end.distance(start).unwrap_or_else(|| {
-            unreachable!(
-                "`T: Numerated` impl error: for any s: T, e: T, e ≥ s ⇔ e.distance(s) == Some(_)"
-            )
-        });
+    fn try_from((start, end): (S, E)) -> Result<Self, Self::Error> {
+        let start: T::Bound = start.into();
+        let end: T::Bound = end.into();
 
-        distance.checked_add(&T::N::one())
+        match (start.unbound(), end.unbound()) {
+            (None, None) => Err(TryFromRangeError::EmptyRange),
+            (None, Some(_)) => Err(TryFromRangeError::IncorrectRange),
+            (start, None) => Self::try_from(start..).map_err(|_| TryFromRangeError::EmptyRange),
+            (Some(start), Some(end)) => Self::try_from_range(start..end),
```

Comment:
We need to make difference between constructors and internal representation. Internal interval is start..=end, but it can be constructed from different ranges/objects. I choose `try_from(x: Bound, y: Bound)` to be not inclusive, because Bound it is supposed to be something not inclusive in upper bound. 

I have added NOTE comment to clarify this moment